File: src/main/java/itmo/is/project/model/trade/TradePolicy.java
package itmo.is.project.model.trade;

import itmo.is.project.model.resource.Resource;
import jakarta.persistence.*;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotNull;
import lombok.*;

@Entity
@Table(name = "trade_policy")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class TradePolicy {

    @Id
    private Integer resourceId;

    @MapsId
    @OneToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "resource_id", referencedColumnName = "id")
    private Resource resource;

    @NotNull
    @Column(name = "station_sells", nullable = false)
    private Boolean stationSells = false;

    @NotNull
    @Min(0)
    @Column(name = "sell_price", nullable = false)
    private Integer sellPrice = 0;

    @Min(0)
    @Column(name = "sell_limit")
    private Integer sellLimit;

    @NotNull
    @Column(name = "station_buys", nullable = false)
    private Boolean stationBuys = false;

    @NotNull
    @Min(0)
    @Column(name = "purchase_price", nullable = false)
    private Integer purchasePrice = 0;

    @Min(0)
    @Column(name = "purchase_limit")
    private Integer purchaseLimit;

    public void setResource(Resource resource) {
        this.resource = resource;
        this.resourceId = resource.getId();
    }
}

File: src/main/kotlin/itmo/isproject/trade/model/TradePolicy.kt
package itmo.isproject.trade.model

import itmo.isproject.shared.resource.model.Resource
import jakarta.persistence.*
import jakarta.validation.constraints.Min
import jakarta.validation.constraints.NotNull

@Entity
@Table(name = "trade_policy")
class TradePolicy(

    @MapsId
    @OneToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "resource_id", referencedColumnName = "id")
    var resource: Resource? = null,

    @NotNull
    @Column(name = "station_sells", nullable = false)
    var stationSells: Boolean = false,

    @NotNull
    @Min(0)
    @Column(name = "sell_price", nullable = false)
    var sellPrice: Int = 0,

    @Min(0)
    @Column(name = "sell_limit")
    var sellLimit: Int? = null,

    @NotNull
    @Column(name = "station_buys", nullable = false)
    var stationBuys: Boolean = false,

    @NotNull
    @Min(0)
    @Column(name = "purchase_price", nullable = false)
    var purchasePrice: Int = 0,

    @Min(0)
    @Column(name = "purchase_limit")
    var purchaseLimit: Int? = null

) {
    @Id
    var resourceId: Int? = null
        get() = resource?.id
        set(value) {
            field = value
            resource = null
        }

    init {
        resourceId = resource?.id
    }
}
